File: src/main/java/cc/mirukuneko/realtrainmodrenewed/blockentity/SignalStateBlockEntity.java
package cc.mirukuneko.realtrainmodrenewed.blockentity;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities;
import cc.mirukuneko.realtrainmodrenewed.block.SignalStateBlock;
import cc.mirukuneko.realtrainmodrenewed.signal.SignalAspect;
import cc.mirukuneko.realtrainmodrenewed.signal.SignalNetworkSavedData;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.level.storage.ValueInput;
import net.minecraft.world.level.storage.ValueOutput;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;

public class SignalStateBlockEntity extends BlockEntity {
    private int linkedChannel = -1;
    private int aspectId = SignalAspect.STOP.getId();

    public SignalStateBlockEntity(BlockPos pos, BlockState blockState) {
        super(RealTrainModRenewedBlockEntities.SIGNAL_STATE.get(), pos, blockState);
    }

    public static void serverTick(ServerLevel level, BlockPos pos, BlockState state, SignalStateBlockEntity blockEntity) {
        if ((level.getGameTime() + pos.asLong()) % 5L != 0L) {
            return;
        }
        int channel = blockEntity.linkedChannel > 0 ? blockEntity.linkedChannel : blockEntity.findAdjacentSourceSignalChannel();
        SignalAspect aspect = channel > 0 ? SignalNetworkSavedData.get(level).getAspect(channel) : SignalAspect.STOP;
        blockEntity.updateState(level, pos, state, channel, aspect);
    }

    private void updateState(ServerLevel level, BlockPos pos, BlockState state, int channel, SignalAspect aspect) {
        int nextAspectId = aspect == null ? SignalAspect.STOP.getId() : aspect.getId();
        boolean changed = aspectId != nextAspectId;
        aspectId = nextAspectId;
        applyAspectToAttachedSignals(aspect);
        if (!changed) {
            return;
        }
        if (state.hasProperty(SignalStateBlock.ASPECT) && state.getValue(SignalStateBlock.ASPECT) != nextAspectId) {
            level.setBlock(pos, state.setValue(SignalStateBlock.ASPECT, nextAspectId), 3);
        }
        setChanged();
        level.sendBlockUpdated(pos, getBlockState(), getBlockState(), 3);
    }

    private int findAdjacentSourceSignalChannel() {
        if (level == null) {
            return -1;
        }
        for (Direction direction : Direction.values()) {
            if (level.getBlockEntity(worldPosition.relative(direction)) instanceof InstalledObjectBlockEntity installed
                && installed.isSignal()
                && installed.getSignalChannel() > 0) {
                return installed.getSignalChannel();
            }
        }
        return -1;
    }

    private void applyAspectToAttachedSignals(SignalAspect aspect) {
        if (level == null) {
            return;
        }
        for (Direction direction : Direction.values()) {
            BlockEntity blockEntity = level.getBlockEntity(worldPosition.relative(direction));
            if (blockEntity instanceof InstalledObjectBlockEntity installed && installed.isSignal()) {
                installed.setSignalAspect(aspect, true);
            }
        }
    }

    public int getLinkedChannel() {
        return linkedChannel;
    }

    public void setLinkedChannel(int linkedChannel) {
        this.linkedChannel = linkedChannel;
        setChanged();
    }

    @Override
    protected void saveAdditional(ValueOutput tag) {
        super.saveAdditional(tag);
        tag.putInt("LinkedChannel", linkedChannel);
        tag.putInt("AspectId", aspectId);
    }

    @Override
    protected void loadAdditional(ValueInput tag) {
        super.loadAdditional(tag);
        linkedChannel = tag.getIntOr("LinkedChannel", -1);
        aspectId = tag.getIntOr("AspectId", SignalAspect.STOP.getId());
    }

    @Override
    public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
        return saveWithoutMetadata(registries);
    }

    @Override
    public ClientboundBlockEntityDataPacket getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/blockentity/SignalStateBlockEntity.kt
package cc.mirukuneko.realtrainmodrenewed.blockentity

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities
import cc.mirukuneko.realtrainmodrenewed.block.SignalStateBlock
import cc.mirukuneko.realtrainmodrenewed.signal.SignalAspect
import cc.mirukuneko.realtrainmodrenewed.signal.SignalNetworkSavedData
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.storage.ValueInput
import net.minecraft.world.level.storage.ValueOutput
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState

class SignalStateBlockEntity(pos: BlockPos, state: BlockState) :
    BlockEntity(RealTrainModRenewedBlockEntities.SIGNAL_STATE.get(), pos, state) {

    var linkedChannel: Int = -1
    var aspectId: Int = SignalAspect.STOP.id

    companion object {
        @JvmStatic
        fun serverTick(level: ServerLevel, pos: BlockPos, state: BlockState, be: SignalStateBlockEntity) {
            if ((level.gameTime + pos.asLong()) % 5L != 0L) return
            val channel = if (be.linkedChannel > 0) be.linkedChannel else be.findAdjacentSourceSignalChannel()
            val aspect = if (channel > 0) SignalNetworkSavedData.get(level).getAspect(channel) else SignalAspect.STOP
            be.updateState(level, pos, state, channel, aspect)
        }
    }

    private fun updateState(level: ServerLevel, pos: BlockPos, state: BlockState, channel: Int, aspect: SignalAspect) {
        val nextId = aspect.id
        val changed = aspectId != nextId
        aspectId = nextId
        applyAspectToAttachedSignals(aspect)
        if (!changed) return
        if (state.hasProperty(SignalStateBlock.ASPECT) && state.getValue(SignalStateBlock.ASPECT) != nextId)
            level.setBlock(pos, state.setValue(SignalStateBlock.ASPECT, nextId), 3)
        setChanged()
        level.sendBlockUpdated(pos, blockState, blockState, 3)
    }

    private fun findAdjacentSourceSignalChannel(): Int {
        if (level == null) return -1
        for (dir in Direction.entries) {
            val be = level!!.getBlockEntity(worldPosition.relative(dir))
            if (be is InstalledObjectBlockEntity && be.isSignal && be.signalChannel > 0)
                return be.signalChannel
        }
        return -1
    }

    private fun applyAspectToAttachedSignals(aspect: SignalAspect) {
        if (level == null) return
        for (dir in Direction.entries) {
            val be = level!!.getBlockEntity(worldPosition.relative(dir))
            if (be is InstalledObjectBlockEntity && be.isSignal) be.setSignalAspect(aspect, true)
        }
    }

    fun setLinkedChannel(channel: Int) { linkedChannel = channel; setChanged() }

    override fun saveAdditional(tag: ValueOutput) {
        super.saveAdditional(tag)
        tag.putInt("LinkedChannel", linkedChannel)
        tag.putInt("AspectId", aspectId)
    }

    override fun loadAdditional(tag: ValueInput) {
        super.loadAdditional(tag)
        linkedChannel = tag.getIntOr("LinkedChannel", -1)
        aspectId = tag.getIntOr("AspectId", SignalAspect.STOP.id)
    }

    override fun getUpdateTag(registries: HolderLookup.Provider): CompoundTag = saveWithoutMetadata(registries)
    override fun getUpdatePacket(): ClientboundBlockEntityDataPacket = ClientboundBlockEntityDataPacket.create(this)
}
